neurons/validator.py

```diff
@@ -788,29 +932,89 @@ def log_step(
                     f"Validator has completed {self.run_step_count} run steps. Creating a new wandb run."
                 )
                 self.wandb_run.finish()
-                self.new_wandb_run()
+                self._new_wandb_run()
 
             original_format_json = json.dumps(step_log)
             uids = step_log["uids"]
             uid_data = step_log["uid_data"]
 
             # Create a new dictionary with the required format
+            with self.metagraph_lock:
+                block = self.metagraph.block.item()
             graphed_data = {
                 "time": time.time(),
-                "block": self.metagraph.block.item(),
+                "step_competition_id": competition_id,
+                "block": block,
                 "uid_data": {
                     str(uid): uid_data[str(uid)]["average_loss"] for uid in uids
                 },
+                "win_rate_data": {
+                    str(uid): uid_data[str(uid)]["win_rate"] for uid in uids
+                },
+                "win_total_data": {
+                    str(uid): uid_data[str(uid)]["win_total"] for uid in uids
+                },
                 "weight_data": {str(uid): self.weights[uid].item() for uid in uids},
-                "load_model_perf_log": load_model_perf_str,
-                "compute_model_perf_log": compute_loss_perf_str,
+                "competition_id": {
+                    str(uid): uid_to_competition_id[uid]
+                    for uid in uids
+                    if uid_to_competition_id[uid] is not None
+                },
+                "load_model_perf": {
+                    "min": load_model_perf.min(),
+                    "median": load_model_perf.median(),
+                    "max": load_model_perf.max(),
+                    "P90": load_model_perf.percentile(90),
+                },
+                "compute_model_perf": {
+                    "min": compute_loss_perf.min(),
+                    "median": compute_loss_perf.median(),
+                    "max": compute_loss_perf.max(),
+                    "P90": compute_loss_perf.percentile(90),
+                },
             }
             bt.logging.trace("Logging to Wandb")
             self.wandb_run.log(
                 {**graphed_data, "original_format_json": original_format_json},
                 step=self.global_step,
             )
 
+    def _get_dataloader_class(
+            self, competition_id: int
+    ) -> IterableDataset:
+        """
+        Decide which dataset loader class to use, using the competition ID.
+        """
+        dataset_name = constants.DATASET_BY_COMPETITION_ID[competition_id]
```

Comment:
You could just use the dataset loader directly in this map to avoid this extra string->class lookup here.

Reply:
I agree